### DISCUSSION TITLE
Concept: Decentralized Reputation Initiation Protocol (DRIP)

### DISCUSSION THREAD

**Discussion by sanity** [freenet/freenet-core#443] [General]
*This is a preliminary discussion for brainstorming purposes, not even close to a complete proposal*.

DRIP (Decentralized Reputation Initiation Protocol) is a mechanism for initializing reputations in a decentralized system similar to the "web of trust" used by PGP and later Freenet, based on the [prisoner's dilemma](https://en.wikipedia.org/wiki/Prisoner%27s_dilemma) problem in game theory. DRIP is designed to encourage cooperation between existing users and discourage [Sybil attacks](https://en.wikipedia.org/wiki/Sybil_attack), while also achieving some other important goals:

* Being more egalitarian, because it benefits trustworthiness rather than wealth
* Avoiding pointlessly burning energy, unlike proof-of-work systems
* DRIP is unique in that it uses "proof-of-trust" as an alternative source of scarcity, rather than relying on economic incentives like proof-of-stake systems or energy consumption like proof-of-work systems. This means that users who are trusted by others in the system have a higher likelihood of being able to create new identities, regardless of their wealth or other resources.

Here's how it works: Two existing users (sponsors) participate in a trust exercise to "sponsor" a new user. They each have a "cooldown period" after sponsoring a user, which is a period of time during which they're unable to participate in the trust exercise again. In addition, there is a "co-sponsorship cooldown period" which is a period of time during which any two sponsors are unable to co-sponsor with each other. The co-sponsorship cooldown period is much longer than the per-sponsor cooldown period to limit the rate at which any two sponsors can co-sponsor with each other. Both the per-sponsor cooldown period and the co-sponsorship cooldown period are specified by a global decentralized system, which monitors the rate at which identities are created and adjusts these cooldown periods to limit this rate.

In addition, there is an "inbreeding prevention" mechanism that prevents sponsors who are too close on the "family tree" from co-sponsoring a new user together. This helps to maintain the diversity and robustness of the system.

Each sponsor secretly decides whether to cooperate or defect, and they commit to their decision by exchanging hashes and signatures. Then, they reveal their cooperate/defect decisions and the protocol determines the outcome based on their decisions:

* If both sponsors cooperate, their cooldown periods are of medium duration and they each get the right to create a new user.
* If one sponsor defects and the other doesn't, the one who defects has a short cooldown period and the one who cooperates has a long cooldown period.

**Comment by AshtonSnapp** [comment]
What exactly does the cooperate/defect system exist for in this proposal? What problem does it solve?

Also, as a side note, it reminds me of the Prisoner's Dilemma: cooperating is better for the group, but defecting is better for the individual assuming only one side defects.

**Reply by sanity** [reply]
Yes, it's based on the prisoners dilemma. The idea is to provide a decentralized way to create identities, but make it hard enough to avoid people creating lots of puppet identities for nefarious reasons (ie. Sybil attack). Does that help clarify?

**Reply by AshtonSnapp** [reply]
That _does_ help clarify, thank you!

One question then: what should happen if both sponsors decide to defect? I'm assuming this is a possibility since, like in the Prisoner's Dilemma, they don't know what the other picked until after they both make their decisions. Would it result in the sponsorship failing?

**Reply by sanity** [reply]
That's a good question, and the answer will depend on exactly how we want the incentives to work.

One possibility if both sponsors defect is that both are punished with a long cooldown period, and then perhaps also with the sponsorship failing.

A higher cost for mutual defection increases the trust required for sponsorship.

I also just realized I forgot to mention that any two given sponsors can only co-sponsor together a restricted number of times - perhaps just once at one extreme, or perhaps with a co-sponsorship cooldown period much longer than the normal cooldown period. Will edit original.

**Comment by TheRook** [comment]
One concern that I have is that any difference in the data set than this would create a fault where some set of network would be seen as untrustworthy.  Although I agree that if you are out of date you shouldn't be considered as a good resource - the concern is that if two sets of nodes were separated for long enough they would create a fork.

Blockchains solve this kind of conflict by choosing the longest chain, but if is essentially free to make new blocks and free to add new clients - then the idea of "trust" could also be artificially inflated on demand.   An adversary is free to spin up an unlimited number of nodes all of which trust each-other implicitly and therefore would appear to be a more trustworthy source. 

Proof-of-stake is nice in that the coins can be taken from a node who is caught in a lie.  Having some amount of centralization is nice too because you have a basis of an honest dataset which means that the trust exercise can be definitive because you have a ground truth to reference.  My gut feeling is that you still need one of these two methods.

**Reply by TheRook** [reply]
Another solution would be to make it expensive to create new accounts.  For example, lets say that any node that uses the DRIP protocol needs to have a public key with a prefix of 7 letter "f"s, so it is easy for all other nodes to see that this data source spent computational effort to make this new identity and so the node stands to lose more by being throttled or banned. If it is free to make new accounts, then a cooldown can be bypassed.

This is the solution from Microsoft's HashCash paper, and doesn't rely on a blockchain or currency, so it would still be free.

**Reply by sanity** [reply]
Hmm, isn't finding the 7-letter prefix another kind of proof of work?

**Reply by TheRook** [reply]
> Hmm, isn't finding the 7-letter prefix another kind of proof of work?

Did you have a solution?  Or if a one time proof-of-work identity the only standing solution here?

**Reply by sanity** [reply]
I'm still not sure I understand the problem you're addressing. The purpose of DRIP/PoT to make it expensive to create new accounts because it requires finding a co-sponsor that you trust, that isn't someone you've co-sponsored with in the recent past. I don't see why PoW would be required (and the point is to avoid PoW).

Perhaps if you described an example of the threat scenario you're concerned about?

**Reply by TheRook** [reply]
> I'm still not sure I understand the problem you're addressing. The purpose of DRIP/PoT to make it expensive to create new accounts because it requires finding a co-sponsor that you trust, that isn't someone you've co-sponsored with in the recent past. I don't see why PoW would be required (and the point is to avoid PoW).
> 
> Perhaps if you described an example of the threat scenario you're concerned about?

Edges between nodes are as easy to forge as the nodes themselves.  We see these types of attacking in the real world, what you are describing is no different from law enforcement going undercover and infiltrating an organization with someone vouching for them.  If it is trivial to make new accounts, then a government like china can boot up more freenet nodes than there are currently real nodes, and these nodes can fake their own metrics to forge sponsors.  This resource is not scarce.

One way to limit fake interaction and fake accounts is to make it more difficult to generate accounts, PoS does this for other networks - but clearly we don't want an economic driver so our options are limited. But, this shouldn't rule out HashCash-like solutions because it's still free.

**Reply by sanity** [reply]
Seems like we're talking past each other.

The purpose of DRIP is to ensure it isn't trivial to make new accounts, and particularly difficult to create a large number of new accounts (due to cooldown and inbreeding prevention). 

It's intended to be an alternative to burning energy (PoW) or staking wealth (PoS), so adding PoW or PoS to it would defeat the purpose.

Is it not clear how DRIP would achieve this without PoW or PoS?

**Reply by TheRook** [reply]
You are approaching one of the most difficult problems in modern cryptography and acting like I should be already aware that you solved the problem.  I don't see any rigor here, and pretending like you don't need to prove your solution to the community isn't in the best interest of any solution or any community.   I am on the top 10 of all time for security and cryptography on stackoverflow, if I can't make heads or tails of your solution -  then who will?

https://stackoverflow.com/users/183528/rook

**Reply by ple1n** [reply]
PoW/HashCash is too inefficient and stupid. And this is not a cryptographic problem (at least not directly solved by it)

**Reply by TheRook** [reply]
> PoW/HashCash is too inefficient and stupid. And this is not a cryptographic problem (at least not directly solved by it)

PoW isn't great, but at least it works, is peer reviewed and over a trillion dollars in value is riding on this "stupid" solution.  Sybil attacks are best solved with cryptography, and one might say can only be solved with cryptography.

**Reply by ple1n** [reply]
It has no use beyond cyber arms race. No one other than miners would participate in this non-sense. Yes it it totally okay to use it for cryptocurrency arms race, but even eth is transistioning now. In locutus it makes absolutely no sense, because mobile users exist and miners' hardware outperform regular PCs by orders of magnitude. 

And 'a trillion dollar' isn't a valid argument. People invent stupid games and people participate in it and that's it. We have other arms races already, nuclear etc. Furthermore, reasonable solutions do exist which I have mentioned (ZKP + ML for biometric proof of personhood) and web of trust.

**Reply by sanity** [reply]
@TheRook 

> You are approaching one of the most difficult problems in modern cryptography and acting like I should already know that you solved the problem. 

I expect nothing other than that if something is unclear you seek clarification before jumping to conclusions. This is an early brainstorming discussion, not a finished proposal.

>  I don't see any rigor here

That's why the title starts with the word "Concept", rigor comes later - I'm just trying to figure out of this approach is worth pursuing.

> and pretending like you don't need to prove your solution to the community isn't in the best interest of any solution or any community

I didn't "pretend" anything of the sort.

> I am on the top 10 of all time for security and cryptography on stackoverflow, if I can't make heads or tails of your solution - then who will?

I'm sure you're very smart, but when you see something you don't understand, you should ask for clarification rather than assuming the worst.

**Comment by TheRook** [comment]
Yes, that is correct.  One that shows that a node is serious about their
identity.

**Comment by ple1n** [comment]
Hmm you are still thinking about this. Apparently, the ultimate goal is to solve sybil attacks. The 'word' sybil presupposes the existence of real identities, which by definition are tied to real humans. For ensuring real identities I suggest Proof of Personhood. Down this rabbithole I found ZKP with ML which I talked about. PoP is not the complete answer tho. 

Even with complete real humans, they can still be bribed/hired to attack the network. And to reach the end goal, PoP is not even a must. We can view each identity as an individual actor, and it doesn't need to be uniquely tied to a human.  An ideal actor of Locutus network executes the protocol 'correctly' which is hard to define but usually means 'lack of malicious human intervention'. We can therefore call it good-will of an actor. An actor can be controlled by a human or not, and a human doesn't necessarily have good-will. What is 'good-will' is defined by the network. 

I would point out that the meaning of 'trust' kinda shifted. It became a means to enforce will, which is not a bad thing. 

It's possible to create an egalitarian blockchain with PoP identities and democratic block production (voting with random sampling).

**Reply by ple1n** [reply]
I also suggest not thinking too much about web of trust. It would work but it's only a means and is not going to yield anything magical

**Reply by TheRook** [reply]
Without a ground truth or central authority, the loci is subjective.  In a world where we trust voters, but voters can be created on demand than the votes themselves are also subjective because the electorate is arbitrary.

**Reply by sanity** [reply]
Is there a good, decentralized solution for proof-of-personhood? That seems more like outsourcing the problem rather than solving it.

**Reply by ple1n** [reply]
@TheRook 

what is 'loci' ? I am not sure what you meant

> voters can be created on demand

PoP would prevent that.

**Comment by TheRook** [comment]
I am not sure what you are talking about. As of this post, this project
does not meet my peer review, I do not recommend this protocol because it
does not improve security.

**Reply by sanity** [reply]
🤔 This is a preliminary concept discussion - we're not seeking anyone's recommendation at this stage.

**Reply by TheRook** [reply]
I really hoped there was a solution here, i'm so disappointed  at how none of you are taking this seriously.

**Reply by sanity** [reply]
@TheRook Get over yourself.

**Comment by TheRook** [comment]
Ok then, well difficult problems aren't for everyone.

**Reply by ple1n** [reply]
You are not solving it either. What have you proposed ? PoW ? It was considered before the project started.

**Reply by TheRook** [reply]
I am all ears, you can send me a white paper.  Please please please let me know if you have a way to make it difficult to make millions of accounts on demand.   So far i am convinced that you have absolutely no solution what so ever to any real world problem.

**Reply by sanity** [reply]
@TheRook You seem to think it's our job to impress you, take a break and come back when you're ready to contribute constructively.

**Comment by TheRook** [comment]
Making it more difficult to produce accounts with a one time cost will
limit abuse.   No one is suggesting a mining arms race, just making it
somewhat less than trivial to inflate your reputation.
